File: backend/app/api/v1/geocode.py
from dataclasses import asdict

from fastapi import APIRouter, HTTPException, Query

from app.schemas.geocode import (
    GeocodeAutocompleteResponse,
    GeocodeCandidate,
    GeocodeRequest,
    GeocodeResponse,
    GeocodeReverseResponse,
)
from app.services.geocode_providers import GeocodeProviderError
from app.services.geocoding import geocode_autocomplete, geocode_place, geocode_reverse

router = APIRouter(prefix="/geocode", tags=["geocode"])


@router.get("/autocomplete", response_model=GeocodeAutocompleteResponse)
def geocode_autocomplete_endpoint(
    q: str = Query(..., min_length=1, max_length=200),
    city: str = Query("", max_length=100),
    limit: int = Query(5, ge=1, le=10),
) -> GeocodeAutocompleteResponse:
    try:
        outcome = geocode_autocomplete(q, city, limit=limit)
    except GeocodeProviderError as e:
        raise HTTPException(status_code=502, detail=str(e)) from e

    return GeocodeAutocompleteResponse(
        results=[GeocodeCandidate(**asdict(h)) for h in outcome.results],
        provider=outcome.provider,
        warnings=outcome.warnings,
    )


@router.post("/search", response_model=GeocodeResponse)
def geocode_search(body: GeocodeRequest) -> GeocodeResponse:
    try:
        result = geocode_place(body.query, body.destination or "")
    except GeocodeProviderError as e:
        raise HTTPException(status_code=502, detail=str(e)) from e
    if not result:
        raise HTTPException(status_code=404, detail="未找到坐标")
    return GeocodeResponse(**result)


@router.get("/reverse", response_model=GeocodeReverseResponse)
def geocode_reverse_endpoint(
    lat: float = Query(..., ge=-90, le=90),
    lng: float = Query(..., ge=-180, le=180),
) -> GeocodeReverseResponse:
    result = geocode_reverse(lat, lng)
    if not result:
        raise HTTPException(status_code=404, detail="逆地理编码失败")
    return GeocodeReverseResponse(**result)

File: frontend/src/api/geocode.ts
export interface GeocodeResult {
  lat: number;
  lng: number;
  address: string;
  coord_confidence: string;
  coord_source: string;
}

export interface GeocodeCandidate {
  name: string;
  address: string;
  lat: number;
  lng: number;
  place_id: string;
  coord_source: string;
}

export interface GeocodeAutocompleteResponse {
  results: GeocodeCandidate[];
  provider?: string | null;
  warnings?: string[];
}

export class GeocodeApiError extends Error {
  status: number;

  constructor(message: string, status: number) {
    super(message);
    this.name = 'GeocodeApiError';
    this.status = status;
  }
}

const apiBase = () => import.meta.env.VITE_API_BASE ?? '';

export async function geocodeSearch(query: string, destination: string): Promise<GeocodeResult> {
  const res = await fetch(`${apiBase()}/api/v1/geocode/search`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ query, destination }),
  });
  if (!res.ok) {
    let detail = '地理编码失败';
    try {
      const body = (await res.json()) as { detail?: string };
      if (body.detail) detail = body.detail;
    } catch {
      /* ignore */
    }
    throw new GeocodeApiError(detail, res.status);
  }
  return res.json() as Promise<GeocodeResult>;
}

export async function geocodeAutocomplete(
  query: string,
  destination: string,
  limit = 5,
): Promise<GeocodeAutocompleteResponse> {
  const params = new URLSearchParams({ q: query, limit: String(limit) });
  if (destination.trim()) params.set('city', destination.trim());
  const res = await fetch(`${apiBase()}/api/v1/geocode/autocomplete?${params}`);
  if (!res.ok) {
    let detail = '地点联想失败';
    try {
      const body = (await res.json()) as { detail?: string };
      if (body.detail) detail = body.detail;
    } catch {
      /* ignore */
    }
    throw new GeocodeApiError(detail, res.status);
  }
  const data = (await res.json()) as GeocodeAutocompleteResponse;
  return {
    results: data.results ?? [],
    provider: data.provider ?? null,
    warnings: data.warnings ?? [],
  };
}

export interface GeocodeReverseResult {
  name: string;
  address: string;
  lat: number;
  lng: number;
  coord_source: string;
}

export async function geocodeReverse(lat: number, lng: number): Promise<GeocodeReverseResult | null> {
  const params = new URLSearchParams({
    lat: String(lat),
    lng: String(lng),
  });
  const res = await fetch(`${apiBase()}/api/v1/geocode/reverse?${params}`);
  if (!res.ok) return null;
  return res.json() as Promise<GeocodeReverseResult>;
}
